### PR TITLE
Add e2e and integration tests for Service.spec.trafficDistribution

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
#### What type of PR is this?

(tests for a feature)

/kind feature 

#### What this PR does / why we need it:

Add e2e tests for the changes made in https://github.com/kubernetes/kubernetes/pull/123487 regarding `Service.spec.trafficDistribution`

#### Special notes for your reviewer:

I'm trying to follow the latest guidelines about e2e tests from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md (which is why you may see the usage of non-familiar ginkgo/gomega constructs)

We will also need to get https://github.com/kubernetes/test-infra/pull/32190 in -- It will enable the alpha gate for this feature in the respective test environments.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/4444-service-routing-preference
```

/cc @robscott 
/sig network
/priority important-soon

### CODE REVIEW

**Comment by gauravkghildiyal** [comment]
/hold

Will wait for https://github.com/kubernetes/test-infra/pull/32190 to merge to be able to test with presubmits. Though I have locally tested it successfully

File: test/e2e/network/traffic_distribution.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+				}
+			}
+
+			ginkgo.By(fmt.Sprintf("creating 1 pod each in 2 zones %v (out of the total 3 zones)", zones[:2]))
+			zoneForServingPod := make(map[string]string)
+			var servingPods []*corev1.Pod
+			servingPodLabels := map[string]string{"app": f.UniqueName}
+			for _, zone := range zones[:2] {
+				pod := e2epod.NewAgnhostPod(f.Namespace.Name, "serving-pod-in-"+zone, nil, nil, nil, "serve-hostname")
+				pod.Spec.NodeName = nodeForZone[zone]
```
Comment (aojea) [inline test/e2e/network/traffic_distribution.go:151]: we don't want to shortcut the scheduler as it has undesired consequences and creates flakiness and unstability on the test suite, use

```
		nodeSelection := e2epod.NodeSelection{Name: clientNodeInfo.name}
		e2epod.SetNodeSelection(&pod.Spec, nodeSelection)
```

File: test/e2e/network/traffic_distribution.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+			ginkgo.DeferCleanup(framework.IgnoreNotFound(c.CoreV1().Services(f.Namespace.Name).Delete), svc.GetName(), metav1.DeleteOptions{})
+
+			ginkgo.By("ensuring EndpointSlice for service have correct same-zone hints")
+			gomega.Eventually(ctx, endpointSlicesForService(svc.GetName())).WithPolling(5 * time.Second).WithTimeout(e2eservice.ServiceEndpointsTimeout).Should(endpointSlicesHaveSameZoneHints)
+
+			ginkgo.By("keeping traffic within the same zone as the client, when serving pods exist in the same zone")
+
+			createClientPod := func(ctx context.Context, zone string) *corev1.Pod {
+				pod := e2epod.NewAgnhostPod(f.Namespace.Name, "client-pod-in-"+zone, nil, nil, nil)
+				pod.Spec.NodeName = nodeForZone[zone]
```
Comment (aojea) [inline test/e2e/network/traffic_distribution.go:185]: ditto https://github.com/kubernetes/kubernetes/pull/123812/files#r1517231074

File: test/e2e/network/traffic_distribution.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+				return false, nil
+			})
+
+			gomega.Eventually(ctx, requestsFromClient(clientPod)).WithPolling(5 * time.Second).WithTimeout(e2eservice.KubeProxyLagTimeout).Should(requestsSucceedByReachingAnyServingPod)
+
+		})
+
+	})
+
+	ginkgo.When("Service initially has trafficDistribution=PreferClose and transitions to annotation service.kubernetes.io/topology-mode=Auto and back", func() {
```
Comment (aojea) [inline test/e2e/network/traffic_distribution.go:269]: This is an integration test, check https://github.com/kubernetes/kubernetes/tree/master/test/integration/endpointslice or https://github.com/kubernetes/kubernetes/blob/master/test/integration/service/service_test.go as examples

File: test/e2e/network/traffic_distribution.go (hunk 1 above)
Comment (gauravkghildiyal) [reply test/e2e/network/traffic_distribution.go:151]: Oh interesting.

Does this mean the documentation at https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes/#create-a-pod-that-gets-scheduled-to-specific-node is more of a hack? Or maybe it is outdated after we encountered issues with that approach. Just curious.

Sure I'll make the change.

Comment (aojea) [reply test/e2e/network/traffic_distribution.go:151]: is a hack in the sense that you override the scheduler, you can do it, but then you deal with the consequences of doing manual scheduling

Comment (gauravkghildiyal) [reply test/e2e/network/traffic_distribution.go]: Gotcha! Changed

File: test/e2e/network/traffic_distribution.go (hunk 3 above)
Comment (gauravkghildiyal) [reply test/e2e/network/traffic_distribution.go]: Thanks. Rewrote as an integration-test.

File: test/integration/service/service_test.go [hunk 4]
```diff
@@ -406,3 +416,283 @@ func Test_ServiceClusterIPSelector(t *testing.T) {
...
+
+		for _, slice := range slices.Items {
+			for _, endpoint := range slice.Endpoints {
+				var ip string
+				if len(endpoint.Addresses) > 0 {
+					ip = endpoint.Addresses[0]
+				}
+				if endpoint.Hints != nil && len(endpoint.Hints.ForZones) != 0 {
+					fmt.Fprintf(logsBuffer, "endpoint with ip %v has hint %+v, want no hint\n", ip, endpoint.Hints)
+					return false, nil
```
Comment (aojea) [inline test/integration/service/service_test.go:575]: this seems something that should error , no? why we should have a hint on a recently created slice?

File: test/integration/service/service_test.go [hunk 5]
```diff
@@ -406,3 +416,283 @@ func Test_ServiceClusterIPSelector(t *testing.T) {
...
+				if len(endpoint.Addresses) > 0 {
+					ip = endpoint.Addresses[0]
+				}
+				var zone string
+				if endpoint.Zone != nil {
+					zone = *endpoint.Zone
+				}
+				if endpoint.Hints == nil || len(endpoint.Hints.ForZones) != 1 || endpoint.Hints.ForZones[0].Name != zone {
+					fmt.Fprintf(logsBuffer, "endpoint with ip %v does not have the correct hint, want hint for zone %q\n", ip, zone)
+					return false, nil
```
Comment (aojea) [inline test/integration/service/service_test.go:626]: error too, sme comment as before this does not seem something we should retry

File: test/integration/service/service_test.go (hunk 4 above)
Comment (aojea) [reply test/integration/service/service_test.go:575]: I see, so you reuse this function later

File: test/e2e/network/traffic_distribution.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+		c = f.ClientSet
+		e2eskipper.SkipUnlessMultizone(ctx, c)
+	})
+
+	////////////////////////////////////////////////////////////////////////////
+	// Helper functions
+	////////////////////////////////////////////////////////////////////////////
+
+	// endpointSlicesForService returns a helper function to be used with
+	// gomega.Eventually(...). It fetches the EndpointSlcies for the given
```
Comment (aojea) [inline test/e2e/network/traffic_distribution.go:61]:
```suggestion
	// gomega.Eventually(...). It fetches the EndpointSlices for the given
```

File: test/e2e/network/traffic_distribution.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+			servingPodLabels := map[string]string{"app": f.UniqueName}
+			for _, zone := range zones[:2] {
+				pod := e2epod.NewAgnhostPod(f.Namespace.Name, "serving-pod-in-"+zone, nil, nil, nil, "serve-hostname")
+				nodeSelection := e2epod.NodeSelection{Name: nodeForZone[zone]}
+				e2epod.SetNodeSelection(&pod.Spec, nodeSelection)
+				pod.Labels = servingPodLabels
+
+				servingPods = append(servingPods, pod)
+				zoneForServingPod[pod.Name] = zone
+				ginkgo.DeferCleanup(framework.IgnoreNotFound(c.CoreV1().Pods(f.Namespace.Name).Delete), pod.GetName(), metav1.DeleteOptions{})
```
Comment (aojea) [inline test/e2e/network/traffic_distribution.go:155]: this is not really needed FYI, the framework create namespaces for each test and the deletion of the namesapced cascade delete all the objects inside

File: test/e2e/network/traffic_distribution.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+			}
+			e2epod.NewPodClient(f).CreateBatch(ctx, servingPods)
```
Comment (aojea) [inline test/e2e/network/traffic_distribution.go:157]: didn't know about this 😄

File: test/e2e/network/traffic_distribution.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+			for _, clientZone := range zones[:2] {
+				framework.Logf("creating a client pod for probing the service from zone=%q which also has a serving pod", clientZone)
+				clientPod := createClientPod(ctx, clientZone)
+
+				framework.Logf("ensuring that requests from clientPod=%q on zone=%q stay in the same zone", clientPod.Name, clientZone)
+
+				requestsSucceedAndStayInSameZone := gcustom.MakeMatcher(func(reverseChronologicalLogLines []string) (bool, error) {
+					logLines := reverseChronologicalLogLines
+					if len(logLines) < 20 {
+						return false, fmt.Errorf("got %d log lines, waiting for at least 10", len(logLines))
```
Comment (aojea) [inline test/e2e/network/traffic_distribution.go:204]:
```suggestion
						return false, fmt.Errorf("got %d log lines, waiting for at least 20", len(logLines))
```

shoulf you not try to retry here until you have enough log lines?

File: test/integration/service/service_test.go (hunk 4 above)
Comment (gauravkghildiyal) [reply test/integration/service/service_test.go:436]: Yep that is the case.

File: test/integration/service/service_test.go (hunk 5 above)
Comment (gauravkghildiyal) [reply test/integration/service/service_test.go:487]: I guess this is also same as https://github.com/kubernetes/kubernetes/pull/123812#discussion_r1522146298

**Comment by gauravkghildiyal** [comment]
/unhold 

(The dependent PR https://github.com/kubernetes/test-infra/pull/32190 is now merged)

File: test/e2e/network/traffic_distribution.go (hunk 9 above)
Comment (aojea) [reply test/e2e/network/traffic_distribution.go]: I don't like this cryptic libraries :( , but I leave implementers to choose their own poison 😄

File: test/e2e/network/traffic_distribution.go (hunk 7 above)
Comment (aojea) [reply test/e2e/network/traffic_distribution.go:164]: good point

File: test/integration/service/service_test.go (hunk 5 above)
Comment (aojea) [reply test/integration/service/service_test.go:487]: yep, forget to update the comment sorry

File: test/e2e/network/traffic_distribution.go (hunk 9 above)
Comment (gauravkghildiyal) [reply test/e2e/network/traffic_distribution.go]: Hahaha!

I totally get you :)

**Comment by gauravkghildiyal** [comment]
Test seems to be passing

**Comment by aojea** [comment]
/lgtm
/approve 
thanks

**Pull request merged**